gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc

```diff
@@ -40,10 +42,60 @@ namespace {
 unsigned int file_counter = 0;
 }
 
+enum class annotator_type { exegesis, fast };
+
+bool AbslParseFlag(absl::string_view text, annotator_type* type,
+                   std::string* error) {
+  if (text == "exegesis") {
+    *type = annotator_type::exegesis;
+    return true;
+  } else if (text == "fast") {
+    *type = annotator_type::fast;
+    return true;
+  }
+
+  *error = "unknown annotator type";
+  return false;
+}
+
+std::string AbslUnparseFlag(annotator_type type) {
+  switch (type) {
+    case annotator_type::exegesis:
+      return "exegesis";
+    case annotator_type::fast:
+      return "fast";
+  }
+}
```

Comment:
Optional: I'd define those as something like
```
constexpr std::pair<AnnotatorType, std::string_view> kAnnotatorTypeNames = {
    {AnnotatorType::kExegesis, "exegesis"},
    {AnnotatorType::kFast, "fast"}
};
```
then parsing and unparsing can both be done with a single for loop, and if we ever add more values, only one place needs to be changed.

Reply:
Is that supposed to be a `constexpr std::array<std::pair<AnnotatorType, std::string_view>,2>`? We don't have `constexpr std::vector` as we aren't using C++20. I've switched to that and it's definitely a lot cleaner.

I used a `__builtin_unreachable` for the `AbslUnparseFlag` as clang complains about not having a return at the end of the function even though that should never be reached. Not sure what the typical practice for this is in Google C++.

Reply:
Oops, sorry, I forgot the array brackets. What I meant was a plain old array
```
constexpr std::pair<AnnotatorType, std::string_view> kAnnotatorTypeNames[] = {
    {AnnotatorType::kExegesis, "exegesis"},
    {AnnotatorType::kFast, "fast"}
};
```
You can get STL-compatible iterators for it with `std::begin()` and `std::end()`, so all the `std::find()` would work. Nice thing about the plain old array is that it doesn't need to have the size specified.

But `std::array<...>` is fine too.

Reply:
Ah, makes sense. I'll switch to a normal array to avoid the extra parameter in the `std::array` template. Any preference on a for loop vs `std::find`?

Reply:
No preference, whatever is simpler/shorter :)